s2and/extentions/clustering_models.py

```diff
@@ -1,47 +1,57 @@
-import json 
+import json
 import torch
 import numpy as np
 from collections import defaultdict
-from typing import Callable, Union, List, Dict
+from os.path import join
+from typing import Tuple, Any, Union, List, Dict
 from tqdm import tqdm
 from joblib import load
 from s2and.data import ANDData
 from sklearn.cluster import DBSCAN, AgglomerativeClustering
 from s2and.extentions.utils import load_signatures
+from s2and.extentions.featurization.operations import Registry
+
 
 class Clusterer():
     """
     Responsible for clustering the block_dict given as input
-    to predict method. 
+    to predict method.
     """
 
-    def __init__(self, 
-                combined_classifier : str,
-                dataset_name : str, 
-                featurization_function : Callable, 
-                default_embeddings: bool = True,
-                embeddings_path: Union[str, None] = None,
-                clusterer: str = 'dbscan') -> None:
+    def __init__(
+        self,
+        combined_classifier: str,
+        dataset_name: str,
+        features: List[Dict[str, str]],
+        embeddings_dir: Union[str, None] = None,
+        clusterer: str = 'dbscan'
+    ) -> None:
 
         self.signatures = load_signatures(dataset_name)
-        self.featurization_function = featurization_function
-        self.default_embeddings = default_embeddings
-        self.embeddings_path = embeddings_path
-        if not default_embeddings:
-            with open(embeddings_path) as f:
+        self.features = features
+        if embeddings_dir is not None:
+            self.embeddings_path = join(embeddings_dir, dataset_name,
+                                        f'{dataset_name}_embeddings.json')
+            with open(self.embeddings_path) as f:
                 self.paper_ids_to_emb = json.load(f)
+        else:
+            self.paper_ids_to_emb = None
 
         self.model = load(combined_classifier)
         self.clusterer_name = clusterer
 
-        if clusterer ==  'dbscan':
+        if clusterer == 'dbscan':
             self.clusterer = DBSCAN(eps=0.4, min_samples=1, metric='precomputed', n_jobs=-1)
         elif clusterer == 'agglomerative':
-            self.clusterer = AgglomerativeClustering(n_clusters=None, affinity='precomputed',
-                             distance_threshold=0.4, linkage='average')
+            self.clusterer = AgglomerativeClustering(
+                n_clusters=None,
+                affinity='precomputed',
+                distance_threshold=0.4,
+                linkage='average'
```

Comment:
you can load these parameters from a config file

Reply:
These are random params just for the sake of initialization. Hyperparam opt follows. A better design would be to give the clusterer object (dbscan or aglomerative) as parameter.